p2p/host/relay/autorelay.go

```diff
@@ -124,28 +139,38 @@ func (ar *AutoRelay) background(ctx context.Context) {
 	}
 }
 
-func (ar *AutoRelay) findRelays(ctx context.Context) bool {
-	if ar.numRelays() >= DesiredRelays {
-		return false
-	}
+func (ar *AutoRelay) findRelays(ctx context.Context, startFindRelaysc <-chan bool, updatec chan<- bool) {
+	running := false
 
-	update := false
-	for retry := 0; retry < 5; retry++ {
-		if retry > 0 {
-			log.Debug("no relays connected; retrying in 30s")
-			select {
-			case <-time.After(30 * time.Second):
-			case <-ctx.Done():
-				return update
-			}
+	checkUpdate := func(ctx context.Context) {
+		if ar.numRelays() >= DesiredRelays {
+			updatec <- false
 		}
 
-		update = ar.findRelaysOnce(ctx) || update
+		update := ar.findRelaysOnce(ctx)
 		if ar.numRelays() > 0 {
-			return update
+			updatec <- update
+			log.Debugf("relay findRelays update %v", update)
+		}
+	}
+
+	for {
+		select {
+		case start := <-startFindRelaysc:
```

Comment:
since `findRelaysOnce` will block this thread for a while, it's very possible that in situations where network connectivity changes (a laptop waking up from sleep, for instance), an ongoing `checkUpdate` on this thread, will mean cause the `startFindRelaysc` channel to fill up. As written, this will cause the background channel to block as well.

Instead, it's maybe worth having the most recent Reachability as a property on the AutoRelay object, and using the channel to signal that "a change has happened". Then, at line 105/107, you can have a default case to ignore if there is already a pending change in the channel, so that the background thread doesn't block when a search is already in progress.